### ISSUE TITLE
Tech discovery: VSO queue and access permissions

### ISSUE THREAD

**Issue by laurjpeterson** [department-of-veterans-affairs/caseflow#6247]
## Background
- Representatives can be affiliated with more than one Veteran Service Organization.
- All VSO representatives should have access to all Veterans who have elected their organization as their POA.
- VSOs currently have access to their Veterans in VACOLS. See Jed's answers to a few VSO access permission questions - https://github.com/department-of-veterans-affairs/dsva-vacols/issues/9

## Proposed initial VSO scope - to be verified
- All VSO users within a given VSO are only given access to appeals where they are the VSO
- All VSO users within a given VSO can:
   - View a single list of appeals where they are the VSO
       - Start with AMA cases only
       - Future state, they will need to see legacy cases too
   - Search for cases and only view results with their own appeals
   - View case details

([VSO Epic mural](https://app.mural.co/t/workqueue2001/m/workqueue2001/1531324988612/3a127a41b4a52fa4b2d36ef969848b6f34693fef))

## Requirements + associated questions
- VSOs cannot have access to Caseflow Reader at this time. So, we will need to create a new CSEM role or function for them. Should we create multiple activity-based functions in CSEM at this time, or create a `VSO Caseflow` function for VSO access?
   - Currently, users have access to queue only if they have access to the Reader CSEM function
   - We've discussed activity-based permissions, and will likely need these in the long run.
   - VSOs would have permission to activities like: `Search cases + view case details`, `Assign back to Board` (when finished with IHP), `View cases associated with VSO` (we aren't supporting work management functionality for VSOs at the outset).
    - It's likely that VSOs will eventually get access to Reader, when we have bandwidth to hide attorney/judge comments, and probably other changes
- Within all of these activities, though, they will need to see a filtered view of cases - cases only for Veterans that have elected the VSOs that they are accredited with.
    - We are planning to get POA from BGS - **how will we use that field to filter VSO views?**
    - Do we need to / should we verify accreditation? This may require us working with another system,  OGC's GCLAWS (I think)
- VSOs do NOT need to be restricted access to the particular issues that a Veteran has elected a different POA for within a given appeal. This is because the VSO already has access to the entire eFolder, so it wouldn't really mean they have restricted access.
    - [Slack thread](https://dsva.slack.com/archives/CA1QBDLEL/p1531839363000090) that discusses this with Dustin Ware and LRP
- If a Veteran has elected a different VSO POA for a single issue (a POA exception), that VSO needs access to see that case. (this is likely a very rare case)

## Resources
- [Queue Discovery from 2017 Mural](https://app.mural.co/t/workqueue2001/m/workqueue2001/1505353215232/e9fee82ce547bd3fc9f1d6cea01947f5b2527920) about how VSOs receive and move cases for which they write IHPs

**Comment by mdbenjam** [comment]
Some initial thoughts:
1) `Do we need to / should we verify accreditation? This may require us working with another system, OGC's GCLAWS (I think)`
   - I don't think so, I believe that information is already in BGS from what I can see. I think they get it from OGC, so we should be able to lean on BGS for that info.
1) The current flow for multiple PoA in VBMS is real lame. They have to manually change between PoAs to give the different orgs/attorneys access to the case in VBMS. This is something we can finally solve with Reader and Queue. But agreed that it's not a super common case and can be postponed.

**Comment by laurjpeterson** [comment]
Mural about VSOs with my initial proposed scope - let's compile questions here: https://app.mural.co/t/workqueue2001/m/workqueue2001/1531324988612/3a127a41b4a52fa4b2d36ef969848b6f34693fef

**Comment by mdbenjam** [comment]
Okay, after doing investigation in BGS, I cannot figure out how to determine the VSO a given VSO employee belongs to, here are my suggestions for how to go forward. This might change as we go forward, but seems like a good first step.

1) Create new CSS roles for each VSO i.e. VSO-VFW, VSO-AL, etc..
2) When we receive appeals, we determine their representative and create a task assigned to the VSO.
3) Each VSO has it's own queue page which lists all of the appeals assigned to them. This might be a whole lot, so we should consider truncating it to the most recent 100 or something.
4) Optional, if we have time we can build a generic assignment tool.
   1) Populate an assignment dropdown with a list of all users in our system with the given VSO role. This can be used to assign a task to that user.
5) We'll need some sort of button to signify that the VSO is done and the task is complete.

For individual attorney POAs not a part of a VSO, we can use the above BGS call to the `findPtcpntId` endpoint in the `SecurityWebService` to get the participant id and then verify that against the participant id of the POA assigned to a given case. But this likely doesn't need to happen until February.

**Comment by mdbenjam** [comment]
Update: Thanks to our VBMS friends we know how to do this now: https://github.com/department-of-veterans-affairs/dsva-vbms/issues/105#issuecomment-406289702.

We can call the `findPOAsByPtcpntId` endpoint in the org service to get the POAs that a given user is associated with.

**Comment by mdbenjam** [comment]
I think this means we just need one role, `VSO`

**Comment by laurjpeterson** [comment]
closing, in favor of our VSO epic and subsequent github issues to implement VSO access permissions
